test/unit/range/iterator_test_template.hpp

```diff
@@ -642,6 +651,69 @@ TYPED_TEST_P(iterator_fixture, compare_geq)
     }
 }
 
+// ---------------------------------------------------------------------------------------------------------------------
+// Contiguous Iterator
+// ---------------------------------------------------------------------------------------------------------------------
+
+TYPED_TEST_P(iterator_fixture, address_test)
+{
+    if constexpr (std::derived_from<typename TestFixture::iterator_tag, std::contiguous_iterator_tag>)
+    {
+        auto begin_it = std::ranges::begin(this->test_range);
+        EXPECT_EQ(std::to_address(begin_it), std::addressof(*begin_it));
+    }
+
+    if constexpr (std::derived_from<typename TestFixture::iterator_tag, std::contiguous_iterator_tag> &&
+                  TestFixture::const_iterable)
+    {
+        auto cbegin_it = std::ranges::cbegin(this->test_range);
+        EXPECT_EQ(std::to_address(cbegin_it), std::addressof(*cbegin_it));
+    }
+}
+
+template <typename it_begin_t, typename it_sentinel_t>
+inline void address_difference_test(it_begin_t && it_begin, it_sentinel_t && it_end)
+{
+    using difference_t = std::iter_difference_t<it_begin_t>;
+
+    for (auto it = it_begin; it != it_end; ++it)
+        EXPECT_EQ(std::to_address(it), std::to_address(it_begin) + static_cast<difference_t>(it - it_begin));
+}
+
+TYPED_TEST_P(iterator_fixture, address_difference)
+{
+    if constexpr (std::derived_from<typename TestFixture::iterator_tag, std::contiguous_iterator_tag>)
+    {
+        address_difference_test(std::ranges::begin(this->test_range), std::ranges::end(this->test_range));
+    }
+
+    if constexpr (std::derived_from<typename TestFixture::iterator_tag, std::contiguous_iterator_tag> &&
+                  TestFixture::const_iterable)
+    {
+        address_difference_test(std::ranges::cbegin(this->test_range), std::ranges::cend(this->test_range));
+    }
+}
+
+TYPED_TEST_P(iterator_fixture, address_to_end)
+{
+    if constexpr (std::derived_from<typename TestFixture::iterator_tag, std::contiguous_iterator_tag>)
+    {
+        using difference_t = std::iter_difference_t<decltype(std::ranges::begin(this->test_range))>;
+        auto it_begin = std::ranges::begin(this->test_range);
+        auto it_end = std::ranges::end(this->test_range);
+        EXPECT_EQ(std::to_address(it_end), std::to_address(it_begin) + static_cast<difference_t>(it_end - it_begin));
+    }
+
+    if constexpr (std::derived_from<typename TestFixture::iterator_tag, std::contiguous_iterator_tag> &&
+                  TestFixture::const_iterable)
+    {
+        using difference_t = std::iter_difference_t<decltype(std::ranges::cbegin(this->test_range))>;
+        auto it_cbegin = std::ranges::cbegin(this->test_range);
+        auto it_cend = std::ranges::cend(this->test_range);
+        EXPECT_EQ(std::to_address(it_cend), std::to_address(it_cbegin) + static_cast<difference_t>(it_cend - it_cbegin));
+    }
+}
```

Comment:
ins't this tested already before implicitly?

Reply:
Oh, Yes you are right. I'll delete the test.

Reply:
I like this as it is now. Because we have in the general the following cases:

1. difference between iterator, iterator
2. difference between iterator, sentinel (and vice versa)
3. difference between sentinel, sentinel

So that means we should also have 3 test cases that check that the addresses are contiguous for all these cases.

1. is done by address_difference and address_test
2. is done by address_of_end and by my suggestion for address_difference
3. is done by address_to_end